--- /dev/null
+++ b/src/main/java/me/earthme/lophine/config/modules/misc/ContainerExpansionConfig.java
@@ -1,0 +_,27 @@
+package me.earthme.lophine.config.modules.misc;
+
+import me.earthme.luminol.config.EnumConfigCategory;
+import me.earthme.luminol.config.IConfigModule;
+import me.earthme.luminol.config.flags.ConfigInfo;
+
+public class ContainerExpansionConfig implements IConfigModule {
+    @ConfigInfo(baseName = "barrel_rows", comments =
+            """
+                    range: 1~6\s""")
+    public static int barrelRows = 3;
+
+    @ConfigInfo(baseName = "enderchest_rows", comments =
+            """
+                    range: 1~6\s""")
+    public static int enderchestRows = 3;
+
+    @Override
+    public EnumConfigCategory getCategory() {
+        return EnumConfigCategory.MISC;
+    }
+
+    @Override
+    public String getBaseName() {
+        return "container_expansion";
+    }
+}
