From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 7 Jul 2025 16:30:54 +0800
Subject: [PATCH] Add config to allow infinite trade


diff --git a/net/minecraft/world/entity/npc/Villager.java b/net/minecraft/world/entity/npc/Villager.java
index 79242037e868a140513f032e74e8594c0c6ff41c..050e89c3e95c7c3ef7ad98f8ef03ab206241a7e2 100644
--- a/net/minecraft/world/entity/npc/Villager.java
+++ b/net/minecraft/world/entity/npc/Villager.java
@@ -490,8 +490,15 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
     }
 
     private void resendOffersToTradingPlayer() {
-        MerchantOffers offers = this.getOffers();
+        // Lophine start - infinite trade
+        MerchantOffers offers = this.getOffers().copy();
         Player tradingPlayer = this.getTradingPlayer();
+        if (fun.bm.lophine.config.modules.function.VillagerConfig.villagerInfiniteTrade) {
+            for (MerchantOffer merchantOffer : offers) {
+                merchantOffer.maxUses = 524288;
+            }
+        }
+        // Lophine end - infinite trade
         if (tradingPlayer != null && !offers.isEmpty()) {
             tradingPlayer.sendMerchantOffers(
                 tradingPlayer.containerMenu.containerId,
diff --git a/net/minecraft/world/item/trading/Merchant.java b/net/minecraft/world/item/trading/Merchant.java
index 6e2fa442ddcd4ef3c4fe789b9ebac9dd409c9de7..0c7318a0ed214204f5493576c8a7b02fbe51c00b 100644
--- a/net/minecraft/world/item/trading/Merchant.java
+++ b/net/minecraft/world/item/trading/Merchant.java
@@ -42,7 +42,14 @@ public interface Merchant {
             new SimpleMenuProvider((containerId, inventory, player1) -> new MerchantMenu(containerId, inventory, this), displayName)
         );
         if (optionalInt.isPresent()) {
-            MerchantOffers offers = this.getOffers();
+            // Lophine start - infinite trade
+            MerchantOffers offers = this.getOffers().copy();
+            if (fun.bm.lophine.config.modules.function.VillagerConfig.villagerInfiniteTrade) {
+                for (MerchantOffer merchantOffer : offers) {
+                    merchantOffer.maxUses = 5242884;
+                }
+            }
+            // Lophine end - infinite trade
             if (!offers.isEmpty()) {
                 player.sendMerchantOffers(optionalInt.getAsInt(), offers, level, this.getVillagerXp(), this.showProgressBar(), this.canRestock());
             }
diff --git a/net/minecraft/world/item/trading/MerchantOffer.java b/net/minecraft/world/item/trading/MerchantOffer.java
index 64c99df8ff305fa28c75dc03fc5ef8c61634ad84..770f37d7158c7b4cec62783025f623a68c1f2457 100644
--- a/net/minecraft/world/item/trading/MerchantOffer.java
+++ b/net/minecraft/world/item/trading/MerchantOffer.java
@@ -196,7 +196,7 @@ public class MerchantOffer {
     }
 
     public boolean isOutOfStock() {
-        return this.uses >= this.maxUses;
+        return (this.uses >= this.maxUses && !fun.bm.lophine.config.modules.function.VillagerConfig.villagerInfiniteTrade); // Lophine - infinite trade
     }
 
     public void setToOutOfStock() {
