From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sat, 14 Jun 2025 05:32:51 +0800
Subject: [PATCH] Better ShulkerBox

You can open shulker box with shift & right click

diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 528d0a111b36e51d8943a013b36c378d442ef2dd..2e03b114dd2b8c0816621c7522fb9654025d7111 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -2450,6 +2450,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.containerMenu.removed(this);
         this.inventoryMenu.transferState(this.containerMenu);
         this.containerMenu = this.inventoryMenu;
+        fun.bm.lophine.utils.ShulkerBoxesUtil.clearMap(this); // Lophine - better shulker box
     }
 
     @Override
diff --git a/net/minecraft/world/entity/player/Inventory.java b/net/minecraft/world/entity/player/Inventory.java
index 53210a7d32241c47c2b1cd698b515726f0a1fd0d..399a7424712471329ca82094839859bbac8ed087 100644
--- a/net/minecraft/world/entity/player/Inventory.java
+++ b/net/minecraft/world/entity/player/Inventory.java
@@ -316,6 +316,11 @@ public class Inventory implements Container, Nameable {
 
         int i = this.getMaxStackLeaves(item) - item.getCount();
         int min = Math.min(count, i);
+        // Lophine start - better shulker box
+        if (fun.bm.lophine.config.modules.misc.ContainerExpansionConfig.betterShulker) {
+            fun.bm.lophine.utils.ShulkerBoxesUtil.inventoryCallBack(player, slot);
+        }
+        // Lophine end - better shulker box
         if (min == 0) {
             return count;
         } else {
@@ -368,6 +373,11 @@ public class Inventory implements Container, Nameable {
                     if (slot >= 0) {
                         this.items.set(slot, stack.copyAndClear());
                         this.items.get(slot).setPopTime(5);
+                        // Lophine start - better shulker box
+                        if (fun.bm.lophine.config.modules.misc.ContainerExpansionConfig.betterShulker) {
+                            fun.bm.lophine.utils.ShulkerBoxesUtil.inventoryCallBack(player, slot);
+                        }
+                        // Lophine end - better shulker box
                         return true;
                     } else if (this.player.hasInfiniteMaterials()) {
                         stack.setCount(0);
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 283991c8da3ff07359c89482a72a7238c2cb7e26..5320f57897d673f3468b2bfecee4768bb00ef783 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -605,11 +605,13 @@ public abstract class Player extends LivingEntity {
     // Paper start - special close for unloaded inventory
     public void closeUnloadedInventory(org.bukkit.event.inventory.InventoryCloseEvent.Reason reason) {
         this.containerMenu = this.inventoryMenu;
+        fun.bm.lophine.utils.ShulkerBoxesUtil.clearMap(this); // Lophine - better shulker box
     }
     // Paper end - special close for unloaded inventory
 
     public void closeContainer() {
         this.containerMenu = this.inventoryMenu;
+        fun.bm.lophine.utils.ShulkerBoxesUtil.clearMap(this); // Lophine - better shulker box
     }
 
     protected void doCloseContainer() {
diff --git a/net/minecraft/world/inventory/AbstractContainerMenu.java b/net/minecraft/world/inventory/AbstractContainerMenu.java
index 379ccae078370b4e8fc909e642f1a70066f70ebf..c28629f8900bf4c51a0daa1a8effd6c405ab1526 100644
--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -658,6 +658,11 @@ public abstract class AbstractContainerMenu {
                 }
             }
         }
+        // Lophine start - better shulker box
+        if (fun.bm.lophine.config.modules.misc.ContainerExpansionConfig.betterShulker) {
+            fun.bm.lophine.utils.ShulkerBoxesUtil.inventoryCallBack(player, slotId - 54);
+        }
+        // Lophine end - better shulker box
     }
 
     private boolean tryItemClickBehaviourOverride(Player player, ClickAction action, Slot slot, ItemStack clickedItem, ItemStack carriedItem) {
diff --git a/net/minecraft/world/inventory/ShulkerBoxMenu.java b/net/minecraft/world/inventory/ShulkerBoxMenu.java
index 903025c659e6a9423224fe65973696405c69ec6a..eb20f83d5aa2c06f60c3941e8a54434737c017c4 100644
--- a/net/minecraft/world/inventory/ShulkerBoxMenu.java
+++ b/net/minecraft/world/inventory/ShulkerBoxMenu.java
@@ -53,7 +53,7 @@ public class ShulkerBoxMenu extends AbstractContainerMenu {
 
     @Override
     public boolean stillValid(Player player) {
-        if (!this.checkReachable) return true; // CraftBukkit
+        if (!this.checkReachable || fun.bm.lophine.utils.ShulkerBoxesUtil.checkIfValid(player)) return true; // CraftBukkit // Lophine - better shulker box
         return this.container.stillValid(player);
     }
 
diff --git a/net/minecraft/world/item/Item.java b/net/minecraft/world/item/Item.java
index 40cb5e6bbcf9dbdb400f1503e86c0804b60e07be..8c279c6ed58f1af451170f9633e3a32f9ab59ee0 100644
--- a/net/minecraft/world/item/Item.java
+++ b/net/minecraft/world/item/Item.java
@@ -187,6 +187,13 @@ public class Item implements FeatureElement, ItemLike {
                     player.startUsingItem(hand);
                     return InteractionResult.CONSUME;
                 } else {
+                    // Lophine start - better shulker box
+                    if (fun.bm.lophine.config.modules.misc.ContainerExpansionConfig.betterShulker
+                            && player.isShiftKeyDown()
+                            && fun.bm.lophine.utils.ShulkerBoxesUtil.checkIfCanOpen(itemInHand)) {
+                        fun.bm.lophine.utils.ShulkerBoxesUtil.openShulkerBox(player, itemInHand, hand);
+                    }
+                    // Lophine end - better shulker box
                     return InteractionResult.PASS;
                 }
             }
diff --git a/net/minecraft/world/item/component/ItemContainerContents.java b/net/minecraft/world/item/component/ItemContainerContents.java
index ecf794f94177fc7b6df483516d920719fbc6fa43..4fda59a4c17009a0009f3d3c13258f4ffcb300c3 100644
--- a/net/minecraft/world/item/component/ItemContainerContents.java
+++ b/net/minecraft/world/item/component/ItemContainerContents.java
@@ -33,7 +33,7 @@ public final class ItemContainerContents implements TooltipProvider {
     public final NonNullList<ItemStack> items;
     private final int hashCode;
 
-    private ItemContainerContents(NonNullList<ItemStack> items) {
+    public ItemContainerContents(NonNullList<ItemStack> items) { // Lophine - better shulker box
         if (items.size() > 256) {
             throw new IllegalArgumentException("Got " + items.size() + " items, but maximum is 256");
         } else {
diff --git a/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java b/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
index ebea67223ce1d350087c73dff0cc3fe6d7b47ca0..efbe87de81ad23d0d248dfa487d20190c6049cf4 100644
--- a/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
@@ -46,6 +46,10 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
     private ShulkerBoxBlockEntity.AnimationStatus animationStatus = ShulkerBoxBlockEntity.AnimationStatus.CLOSED;
     private float progress;
     private float progressOld;
+    // Lophine start - better shulker box
+    public boolean haveRealBlock = true;
+    public int slot;
+    // Lophine end - better shulker box
     @Nullable
     private final DyeColor color;
 
@@ -236,6 +240,15 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
         return Component.translatable("container.shulkerBox");
     }
 
+    // Lophine start - better shulker box
+    public void setItem(int index, ItemStack stack) {
+        super.setItem(index, stack);
+        if (fun.bm.lophine.config.modules.misc.ContainerExpansionConfig.betterShulker && !haveRealBlock) {
+            fun.bm.lophine.utils.ShulkerBoxesUtil.shulkerBoxEntityCallBack(this);
+        }
+    }
+    // Lophine end - better shulker box
+
     @Override
     protected void loadAdditional(ValueInput input) {
         super.loadAdditional(input);
@@ -258,12 +271,12 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
     }
 
     @Override
-    protected NonNullList<ItemStack> getItems() {
+    public NonNullList<ItemStack> getItems() { // Lophine - better shulker box
         return this.itemStacks;
     }
 
     @Override
-    protected void setItems(NonNullList<ItemStack> items) {
+    public void setItems(NonNullList<ItemStack> items) { // Lophine - better shulker box
         this.itemStacks = items;
     }
 
